TMEDIA-773 - Carousel Component without Library

matthewroach opened the pull request:
## Ticket

- [TMEDIA-773](https://arcpublishing.atlassian.net/browse/TMEDIA-773)

## Description

Implementation of a carousel component without using a third party carousel library.

Item of note that may appear different to that of the acceptance criteria of the ticket is the implementation of the buttons - I changed the approach to not use `<Carousel.Button>` sub components to reduce complexity of managing how the events would have to be passed to each button.

## Test Steps

1. Checkout branch - `git checkout TMEDIA-773-carousel`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Check out the button storybook documentation

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

matthewroach committed:
Add initial carousel component

matthewroach committed:
Fix lint issues

matthewroach committed:
Lint spacing conflicts with prettier

matthewroach committed:
Scss updates

JackHowa commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,146 @@
+			<div
+				className="c-carousel__track"
+				style={{ transform: `translate3d(${position}%, 0px, 0px)` }}
+				{...handlers}
```
not sure I see the utility here with the handlers? Maybe I'm missing something from the library

matthewroach replied:
The `handlers` is set on line 47 - making use of a library to give us a consistent interface to swipe events rather than having to implement completely custom touch events

JackHowa replied:
<img width="749" alt="Screen Shot 2022-04-01 at 11 10 35" src="https://user-images.githubusercontent.com/5950956/161301632-229d5462-2fcb-4b52-83bf-e5c3b9a91cdc.png">

maybe I'm missing something here. But I'm not able to swipe in the example. do we need refs or something to initiate this? 

I tried reinstalling node modules

JackHowa replied:
locally on phone still not reliably scrolling. maybe it's just me

https://user-images.githubusercontent.com/5950956/161303721-67b7f4e3-9ae8-4962-ab8c-b6f3d8b0048d.MP4

matthewroach replied:
@JackHowa - it's hard to tell what the issue you are seeing is, as I see the page moving up and down which indicates you trying to scroll vertically - which isn't something that will trigger on the carousel. Then I see the carousel move which I assume is because you swiped left

nschubach replied:
I think `.c-carousel__actions` is getting in the way of the touch events.  I can get it to fairly reliably work if I add `trackTouch: true` to the handlers and swipe either the top or bottom of the carousel.

matthewroach replied:
I've moved the handles to the parent element and I get a more consistent experience now

nschubach replied:
added what fixed touch events for me (tested on my android phone)

nschubach replied:
@JackHowa : See if the issue still exists if you can.  I am unable to reproduce after the latest update.

matthewroach committed:
Update single slide story

matthewroach committed:
Remove docs on viewable prop

The viewable prop is set by the carousel component no set by the consumer of the Carousel.Item component

matthewroach committed:
Update Item to allow a child to be a function

matthewroach committed:
Lint issues

matthewroach committed:
Lint and Prettier conflict each other

matthewroach committed:
Add pass through for all attributes

matthewroach committed:
Lint and Prettier conflict each other

nschubach reviewed (commented):
This might get interesting when we get to an infinite scrolling feature.  Not sure I have a problem here, just curious how that will work when wrapping is a concern considering that this just uses an offset of the slide container to manage visibility.

nschubach commented on src/components/carousel/_index.scss:
```diff
@@ -0,0 +1,54 @@
+	--slide-width: var(--carousel-slide-width, 25%);
+
+	display: grid;
+	grid-template: "carousel";
```
```suggestion
	grid-template-areas: "carousel";
```

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,146 @@
+	/** The text, images or any node that will be displayed within the component */
+	children: PropTypes.node.isRequired,
+	/** A unique identifer for the carousel */
+	id: PropTypes.string.isRequired,
```
Does fusion provide a unique identifier that we could use as a default?

matthewroach replied:
I wouldn't want to rely on Fusion here - the consumer could use a Fusion identifier if they wanted too though

nschubach commented on src/components/carousel/index.test.jsx:
```diff
@@ -0,0 +1,186 @@
+				id="carousel-2"
+				label="Carousel Label"
+				slidesToShow={1}
+				next={{ element: <span>Next</span> }}
+				previous={{ element: <span>Previous</span> }}
```
Could we have done something like: 
```
next={
  <Carousel.Button
    label=""
    icon={
      <Icon name="" width={} height={} />
    }
  >
    Next
  </Carousel.Button>
}
```
instead of the complex object for PropTypes?  Maybe even to the point of enabling standard <Button /> theming.

matthewroach replied:
How do you see the `onClick` event being passed down?

nschubach replied:
`<Carousel.Button onClick={} />`?

nschubach replied:
really, just accepting the `...rest` on the `<Carousel.Button />` and passing to the `<button />` would do the same as you are now.

nschubach reviewed (commented):
Mainly some implementation questions

matthewroach committed:
Update grid-template-areas

JackHowa commented on src/components/carousel/index.stories.mdx:
```diff
@@ -0,0 +1,114 @@
+					</div>
+				)}
+			</Carousel.Item>
+			<Carousel.Item label="Slide 3 of 5">
```
is this out of order with the label? Looks like a duplicate

matthewroach committed:
Update story

matthewroach committed:
Change to Carousel.Button

matthewroach committed:
Updates

nschubach committed:
onClick call for next/prev passed components

nschubach committed:
Resolve to type instead of string name value

nschubach committed:
Add tests for custom onClick events

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,164 @@
+		<DefaultNextButton id={id} onClick={() => nextSlide()} />
+	);
+
+	// previousButton.type === Button ?
```
Since we aren't validating
```suggestion

```

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,164 @@
+		trackMouse: true,
+	});
+
+	// nextButton.type === Button ?
```
Since we aren't validating
```suggestion

```

nschubach commented on src/components/carousel/index.stories.mdx:
```diff
@@ -0,0 +1,133 @@
+			label="Carousel of Images"
+			slidesToShow={1}
+			nextButton={
+				<button type="button" onClick={() => console.log("Nick")}>
```
```suggestion
				<button type="button">
```

matthewroach committed:
Update src/components/carousel/index.jsx

Co-authored-by: nschubach <nschubach@gmail.com>

matthewroach committed:
Update src/components/carousel/index.jsx

Co-authored-by: nschubach <nschubach@gmail.com>

matthewroach committed:
Update src/components/carousel/index.stories.mdx

Co-authored-by: nschubach <nschubach@gmail.com>

matthewroach committed:
Move handles to parent

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,162 @@
+	};
+
+	/* istanbul ignore next */
+	const handlers = useSwipeable({
+		onSwipedLeft: () => nextSlide(),
+		onSwipedRight: () => previousSlide(),
+		preventDefaultTouchmoveEvent: true,
+		trackMouse: true,
+	});
```
Part 1/2 of touch fix:
```suggestion
	const handlers = useSwipeable({
		onSwipedLeft: () => nextSlide(),
		onSwipedRight: () => previousSlide(),
		preventDefaultTouchmoveEvent: true,
		trackMouse: true,
		trackTouch: true,
	});
```

nschubach commented on src/components/carousel/_index.scss:
```diff
@@ -0,0 +1,54 @@
+	&__track {
+		display: flex;
+		transition: transform 0.4s ease-in;
+		will-change: transform;
+
```
Unimportant part of touch fix:
```suggestion
		will-change: transform;
		touch-action: 'pan-y';

```

JackHowa replied:
ohh interesting, haven't seen `pan-y` before

nschubach replied:
Stole it from here... 

https://github.com/FormidableLabs/react-swipeable#how-to-use-touch-action-to-prevent-scrolling

...but now that I'm working with it, it doesn't seem to do anything here. :P

nschubach commented on src/components/carousel/_index.scss:
```diff
@@ -0,0 +1,54 @@
+		justify-content: space-between;
+		place-self: center;
+		width: 100%;
+		z-index: 1;
+
+		@include scss.component-properties("carousel-actions");
+	}
+
+	&__button {
+		@include scss.component-properties("carousel-button");
```
Part 2/2 of touch fix:
```suggestion
		z-index: 1;
		pointer-events: none;

		@include scss.component-properties("carousel-actions");
	}

	&__button {
		pointer-events: auto;

		@include scss.component-properties("carousel-button");
```

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,162 @@
+	/** Number of slides to show in view */
+	slidesToShow: PropTypes.number,
+
+	previousButton: PropTypes.node,
+	nextButton: PropTypes.node,
```
Are we adding comments here by convention for storybook?

matthewroach replied:
Updated

nschubach commented on src/components/carousel/index.stories.mdx:
```diff
@@ -0,0 +1,133 @@
+
+<ArgsTable of={Carousel} />
+
+### Carousel.Item
+
+<ArgsTable of={Item} />
```
```suggestion
### Carousel.Item

<ArgsTable of={Item} />

### Carousel.Button

<ArgsTable of={Button} />
```

matthewroach replied:
Updated

matthewroach committed:
Documentation updates

JackHowa reviewed (approved):
Working as expected for me with swiping now! thanks for the effort on this

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,161 @@
+					nextButton.props.onClick(e);
+				}
+			},
+			className: "c-carousel__button c-carousel__button--next",
```
From Huddle, this `className` should merge `nextButton.props.className`

matthewroach replied:
Updated

nschubach commented on src/components/carousel/index.jsx:
```diff
@@ -0,0 +1,161 @@
+					previousButton.props.onClick(e);
+				}
+			},
+			className: "c-carousel__button c-carousel__button--previous",
```
From Huddle, this `className` should merge `previousButton.props.className`

matthewroach committed:
Pass through className to buttons
